File: backend/app/models/report.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel


class Summary(BaseModel):
    row_count: int
    column_count: int
    total_missing_cells: int
    missing_ratio: float
    duplicate_rows: int
    duplicate_ratio: float


class BasicProfile(BaseModel):
    missing_by_column: dict
    inferred_types: dict
    column_stats: dict


class PiiColumn(BaseModel):
    column: str
    detected_types: List[str]


class ColumnOutlierProfile(BaseModel):
    column: str
    mean: Optional[float] = None
    std: Optional[float] = None
    outlier_count: int
    value_count: int
    outlier_ratio: float
    severity: str


class HistoryPoint(BaseModel):
    timestamp: datetime
    score: float
    missing_ratio: float
    outlier_ratio: float


class HistorySnapshot(BaseModel):
    points: List[HistoryPoint] = []


class Alert(BaseModel):
    level: str
    code: str
    message: str


class InsightItem(BaseModel):
    """
    Lightweight AI-style insight generated from the profiling result.
    """

    category: str  # e.g. "missing", "pii", "schema", "drift", "overall"
    severity: str  # "info" | "warning" | "critical" | "suggestion"
    message: str


class DataQualityReport(BaseModel):
    dataset_name: str
    run_id: str
    timestamp: datetime

    summary: Summary
    basic_profile: BasicProfile

    pii_columns: List[PiiColumn] = []
    pii_column_count: int = 0
    has_pii: bool = False

    columns: List[ColumnOutlierProfile] = []
    total_outliers: int = 0
    total_numeric_values: int = 0
    overall_outlier_ratio: float = 0.0

    # Policy / pipeline gate
    contract_suggestion: dict = {}
    policy_passed: bool = True
    policy_failures: List[dict] = []

    # AutoFix
    autofix_plan: List[dict] = []
    autofix_script: str = ""

    # High-level score & ratios
    missing_ratio: float = 0.0
    outlier_ratio: float = 0.0
    overall_score: float = 0.0

    # History + alerts
    history_snapshot: HistorySnapshot = HistorySnapshot()
    alerts: List[Alert] = []

    # ⭐ NEW: AI-style insights
    insights: List[InsightItem] = []

    # ⭐ NEW: inline contract YAML text for preview / download
    contract_yaml: Optional[str] = None

File: frontend/src/types/report.ts
// frontend/src/types/report.ts

export type AlertLevel = "error" | "warning" | "info";

export interface Alert {
  level: AlertLevel;
  code: string;
  message: string;
}

export interface Summary {
  row_count: number;
  column_count: number;
  total_missing_cells: number;
  missing_ratio: number;
  duplicate_rows?: number;
  duplicate_ratio?: number;
}

export interface BasicProfile {
  missing_by_column?: Record<string, number>;
  inferred_types?: Record<string, string>;
  column_stats?: Record<string, any>;
}

export interface ColumnProfile {
  name?: string;
  column?: string;
  dtype?: string;
  missing_ratio?: number;
  outlier_ratio?: number;
  pii_type?: string | null;
  drift_severity?: string | null;
  psi?: number | null;
  [key: string]: any;
}

export interface PolicyFailure {
  code: string;
  message: string;
}

export interface HistoryPoint {
  timestamp: string;
  overall_score?: number;
  missing_ratio?: number;
  outlier_ratio?: number;
}

export interface HistorySnapshot {
  points: HistoryPoint[];
  [key: string]: unknown;
}

export interface PiiColumn {
  column: string;
  detected_types: string[];
}

export interface ScoreGrade {
  letter: string;
  label: string;
  reason?: string;
}

export interface ScoreBreakdown {
  base?: number;
  penalty_missing?: number;
  penalty_outliers?: number;
  penalty_duplicates?: number;
  penalty_pii?: number;
  penalty_drift?: number;
  total_penalty?: number;
  final_score?: number;
  [key: string]: any;
}

export interface SchemaTypeChange {
  column: string;
  before: string | null;
  after: string | null;
}

export interface SchemaPIIInfo {
  has_pii: boolean;
  pii_types?: string[];
}

export interface SchemaPIIChange {
  column: string;
  before: SchemaPIIInfo;
  after: SchemaPIIInfo;
}

export interface SchemaChanges {
  status: string;
  added_columns: string[];
  removed_columns: string[];
  type_changes: SchemaTypeChange[];
  pii_changes: SchemaPIIChange[];
  is_breaking: boolean;
}

export interface DataQualityReport {
  dataset_name: string;
  run_id: string;
  timestamp: string;

  summary?: Summary;
  basic_profile?: BasicProfile;

  overall_score?: number;
  missing_ratio: number;
  outlier_ratio: number;
  duplicate_ratio?: number;

  has_drift?: boolean;
  psi_severity?: string | null;

  columns: ColumnProfile[];

  pii_columns?: PiiColumn[];
  pii_column_count?: number;
  has_pii?: boolean;

  policy_passed: boolean;
  policy_failures: PolicyFailure[];

  alerts: Alert[];

  autofix_plan?: unknown;
  autofix_script?: string;

  history_snapshot?: HistorySnapshot;

  score_grade?: ScoreGrade;
  score_breakdown?: ScoreBreakdown;

  schema_changes?: SchemaChanges;

  [key: string]: unknown;
}
